### PR TITLE
Addon-docs: Add jest-transform.js to convert .mdx => .js for storyshots

### PR DESCRIPTION
**Pull request by danielknell**
Issue: #7223

## What I did

Wrote a jest transform for mdx files.

## How to test

add the following to your jest config:
```json
{
  "transform": {
    "^.+\\.[tj]sx?$": "babel-jest",
    "^.+\\.mdx?$": "@storybook/addon-docs/jest-transform",
  }
}
```

and run jest in a project with storyshots setup and MDX stories.

might need some additional work for things like vue/etc of which i am clueless.

### CODE REVIEW

File: addons/docs/jest-transform.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const mdx = require('@mdx-js/mdx');
+const babel = require('babel-jest');
+const deasyncPromise = require('deasync-promise');
+const createCompiler = require('./mdx-compiler-plugin');
+
+const compilers = [createCompiler({})];
+
+module.exports = {
+  process(src, filename, config, options) {
+    let result = deasyncPromise(mdx(src, { compilers, filepath: filename }));
```
Comment (ndelangen) [inline addons/docs/jest-transform.js]: I wasn't aware jest transformers had to be synchronous.
Just for future reference, here's why: https://github.com/facebook/jest/issues/5556

(this comment is mainly for future-me)

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
Wow awesome first contribution @danielknell !

@shilman are you looking at merging this?

**Comment by shilman** [comment]
@ndelangen absolutely. the source repo seems to have disappeared, so i'll need to resurrect the changes in a new branch & test them. but this is def going in.

**Comment by danielknell** [comment]
`git fetch origin pull/7330/head:pr7330‘

File: addons/docs/jest-transform.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+    result = `/* @jsx mdx */
+    import React from 'react'
+    import { mdx } from '@mdx-js/react'
+    ${result}
+    `;
+
+    return babel.process(result, filename, config, options);
```
Comment (Hypnosphi) [inline addons/docs/jest-transform.js]: It won't work with custom babel-jest transformer (like the one we [have]() in our own repo),  you may want to borrow the approach from [`injectFileName`](https://github.com/storybookjs/storybook/blob/38d7464d2b2db0b36832b70030ea6c815ac72e7d/addons/storyshots/storyshots-core/injectFileName.js) instead

**Comment by Hypnosphi** [comment]
`jest-transform` sounds too generic, let's rename to `jest-transform-mdx` maybe?

**Comment by aendrew** [comment]
What's blocking this from merging? Is there any way I could help? I've started writing stories in MDX and not being able to snapshot test them is blocking me on a few things.

**Comment by shilman** [comment]
@aendrew This is going into 5.3 and we're in the process of releasing 5.2 right now. We'll work through the 5.3 PRs as soon as that's out. I'll let you know if there's anything that can expedite it after 5.2 is out. Thanks!

**Comment by danielknell** [comment]
As a stop gap if you save the transform file locally in the project and change the reference to a relative path you can use it now, that's what my project is doing.

See the linked issue for details.

**Comment by aendrew** [comment]
Just a note, I'm intermittently running into https://github.com/abbr/deasync/issues/55 when using this code in Node 10.16.2 on OS X 10.14.6, I think it's due to `deasync-promise`.

You can remove the dependency on `deasync-promise` entirely by using [`mdx.sync`](https://mdxjs.com/advanced/api#sync-api).

**Comment by ndelangen** [comment]
replaced with this PR: https://github.com/storybookjs/storybook/pull/8189

Because this PR is without a repo now for some reason.

**Pull request closed without merging**